API.md

```diff
@@ -50,33 +50,82 @@ include the usage value formatted at the top of the message.
-```
-h: {
-    alias: 'help'
-}
-```
 
-* `type`: Available types are: `boolean`, `range`, `number`, `string`, and `help`.  Defaults to `string`.
+    ```js
+    h: {
+        alias: 'help'
+    }
+    ```
+
+* `type`: Available types are: `boolean`, `range`, `number`, `string`, `json`, and `help`.  Defaults to `string`.
 
     `help` is a special type that allows the switch to be executed even though
     other paramters are required. Use case is to display a help message and
     quit. This will bypass all other errors, so be sure to capture it. It
     behaves like a `boolean`.
 
+    The `json` type allows building an object using command line arguments that utilize
+    dot-separated (`.`) paths and JSON values. For example, an object argument named
+    `pet` might be built from `--pet '{ "type": "dog" }' --pet.name Maddie`, resulting in
+    the parsing output `{ pet: { type: 'dog', name: 'Maddie' } }`.  The contents of the
+    flags are deeply merged together in the order they were specified.  Additionally,
+    JSON primitives (i.e. `null`, booleans, and numbers) are treated as strings by default,
+    though this behavior may be controlled with the `parsePrimitives` option documented
+    below.  The following example demonstrates the default behavior:
+
+    ```sh
+    # CLI input
+    create-pet --pet.type kangaroo --pet.legs 2 --pet.mammal true \
+               --pet '{ "name": "Maddie", "type": "dog" }' --pet.legs 4
+    ```
+    ```js
+    // Parsing output
+    { pet: { name: 'Maddie', type: 'dog', legs: '4', mammal: 'true' } }
+    ```
+
 * `multiple` : Boolean to indicate if the same argument can be provided multiple times. If true, the parsed value
-will always be an array of `type`'s. Defaults to `false`.
+will always be an array of `type`'s. Defaults to `false`. Does not apply to `json` type arguments.
 
 * `description`: Description message that will be returned with usage information.
 
 * `require`: Boolean to indicate if the argument is required.  Defaults to `false`
 
 * `default`: A default value to assign to the argument if its not provided as an argument.
 
-* `valid`: A value or array of values that the argument is allowed to equal.
+* `valid`: A value or array of values that the argument is allowed to equal. Does not apply to `json` type arguments.
+
+* `parsePrimitives`: When `true`, arguments of the `json` type will parse JSON primitives rather than treat them as strings.  For example, `--pet.name null` will result in the output `{ pet: { name: 'null' } }` by default.  However, when `parsePrimitives` is `true`, the same input would result in the output `{ pet: { name: null } }`.  The same applies for other JSON primitives too, i.e. booleans and numbers.  When this option is `true`, users may represent string values as JSON in order to avoid ambiguity, e.g. `--pet.name '"null"'`.  It's recommended that applications using this option document the behavior for their users.
```

Comment:
Making this an option means that it will be much safer and there will be less surprises. I don't feel that it properly describes that the value will fallback to the string value if the parsing fails.

Maybe this could have another `parsePrimitives: 'strict'` mode, where there is no fallback?

Reply:
I will update the docs to make it clearer 👍  and I will look into `parsePrimitives: 'strict'`.  If I do implement `'strict'` then it would parse primitives strictly and only allow primitives, otherwise return an error.